clang/lib/CodeGen/ItaniumCXXABI.cpp

```diff
@@ -3565,8 +3565,30 @@ void ItaniumRTTIBuilder::BuildVTablePointer(const Type *Ty) {
   // Check if the alias exists. If it doesn't, then get or create the global.
   if (CGM.getItaniumVTableContext().isRelativeLayout())
     VTable = CGM.getModule().getNamedAlias(VTableName);
-  if (!VTable)
-    VTable = CGM.getModule().getOrInsertGlobal(VTableName, CGM.Int8PtrTy);
+
+  // To generate valid device code SPIR-V module with virtual functions these
+  // global pointers should have global address space, not private as by default
+  auto VTableTy =
+      CGM.getLangOpts().SYCLIsDevice &&
+              CGM.getLangOpts().SYCLAllowVirtualFunctions &&
+              (VTableName == ClassTypeInfo || VTableName == SIClassTypeInfo)
+          ? CGM.Int8Ty->getPointerTo(
+                static_cast<unsigned>(LangAS::opencl_global))
+          : CGM.Int8PtrTy;
+  if (!VTable) {
+    if (VTableTy == CGM.Int8PtrTy) {
+      VTable = CGM.getModule().getOrInsertGlobal(VTableName, VTableTy);
+    } else {
+      VTable = CGM.getModule().getOrInsertGlobal(VTableName, VTableTy, [&] {
```

Comment:
Why do we need lambda here?

Reply:
Because we can't pass required address space to `getOrInsertGlobal`, so I use this lambda to construct the required GV

Reply:
Can we just construct it without the lambda?

Reply:
Probably yes, need to try

Reply:
Turns out no, because this function is called several times, so FE generates several variables like GV.1, GV.2 and so on, - so we need to check if this GV exist. I believe we should use getOrInsertGlobal intended to replace check&create.